ledger/types/balance.go

```diff
@@ -30,211 +22,91 @@ func (b *Balance) UnmarshalBinary(data []byte) error {
-func (b *Balance) Unlock(amount Amount) error {
-	locked, subbed := math.CheckedSubU64ToZero(b.Locked, amount.Amount)
-	balance, err := math.CheckedAddU64(b.Balance, subbed)
-	if err != nil {
-		return err
-	}
-	b.Balance = balance
-	b.Locked = locked
-	return nil
+func (b *Balance) IsZero() bool {
+	return b.Amount == 0
 }
 
-// Supply an Amount from Balance.Balance to Balance.Supplied.
-// Returns an error if not sufficient Balance.Balance.
-func (b *Balance) Supply(amount Amount) error {
-	balance, err := math.CheckedSubU64(b.Balance, amount.Amount)
-	if err != nil {
-		return err
-	}
-	supplied, err := math.CheckedAddU64(b.Supplied, amount.Amount)
-	if err != nil {
-		return err
-	}
-	b.Balance = balance
-	b.Supplied = supplied
-	return nil
+func (b *Balance) IsPositive() bool {
+	return b.Amount > 0
 }
 
-// Withdraw an Amount from Balance.Supplied to Balance.Balance.
-// Returns an error if not sufficient Balance.Supplied.
-func (b *Balance) Withdraw(amount Amount) error {
-	supplied, err := math.CheckedSubU64(b.Supplied, amount.Amount)
-	if err != nil {
-		return err
+// Sub from a balance. Error on negative result, mismatched assets, or invalid assets.
+func (b *Balance) Sub(sub *Balance) (*Balance, error) {
+	if !b.Valid() || !sub.Valid() {
+		return nil, errors.New("balance.Sub: invalid input")
 	}
-	balance, err := math.CheckedAddU64(b.Balance, amount.Amount)
-	if err != nil {
-		return err
+	if sub.Amount == 0 {
+		return b, nil // no-op is safe
 	}
-	b.Balance = balance
-	b.Supplied = supplied
-	return nil
-}
-
-// Borrow an Amount from Leverage module and adds it to Balance.Balance and Balance.Borrowed.
-func (b *Balance) Borrow(amount Amount, isVirtual bool) error {
-	balance, err := math.CheckedAddU64(b.Balance, amount.Amount)
-	if err != nil {
-		return err
+	if b.Asset != sub.Asset {
+		return nil, fmt.Errorf("balance.Sub: mismatched assets %s and %s", b.Asset, sub.Asset)
 	}
-
-	if isVirtual {
-		virtual, err := math.CheckedAddU64(b.Virtual, amount.Amount)
-		if err != nil {
-			return err
-		}
-		b.Virtual = virtual
-	} else {
-		borrowed, err := math.CheckedAddU64(b.Borrowed, amount.Amount)
-		if err != nil {
-			return err
-		}
-		b.Borrowed = borrowed
+	if b.Amount < sub.Amount {
+		return nil, fmt.Errorf("balance.Sub (%s): result would be negative", b.Asset)
 	}
-
-	b.Balance = balance
-
-	return nil
+	return NewBalance(b.Asset, b.Amount-sub.Amount), nil
 }
 
-// Repay an Amount from Balance.Borrowed.
-// Returns an error if not sufficient Balance.Borrowed.
-func (b *Balance) Repay(amount Amount) error {
-	borrowed, err := math.CheckedSubU64(b.Borrowed, amount.Amount)
-	if err != nil {
-		return err
+// SubAmt from a balance. Returns ok if balance was valid and result was not negative.
+func (b *Balance) SubAmt(amt uint64) (*Balance, bool) {
+	if !b.Valid() || b.Amount < amt {
+		return nil, false
 	}
-	b.Borrowed = borrowed
-	return nil
+	return NewBalance(b.Asset, b.Amount-amt), true
 }
```

Comment:
I think it would be safer if you created a `Balance` object and used the `Sub` function here, this way all the logic for `Sub` is incorporated into one place in case it needs to be updated in the future.